core/src/main/java/org/ethereum/beacon/core/types/Bitlist.java

```diff
@@ -0,0 +1,221 @@
+    this.size = size;
+    this.maxSize = maxSize;
+  }
+
+  /** Assumes that size info is encoded in bytes\ */
+  public static Bitlist of(BytesValue bytes, long maxSize) {
+    int size = (bytes.size() - 1) * 8;
+    byte lastByte = bytes.get(bytes.size() - 1);
+    int addon = 0;
+    for (int i = 0; i < 8; ++i) {
+      if (((lastByte >> i) & 1) == 1) {
+        addon = i;
+      }
+    }
+    final BytesValue finalBlank;
+    if (addon == 0) {
+      finalBlank = bytes.slice(0, bytes.size() - 1); // last byte was needed only for a size
+    } else {
+      size = size + addon;
+      MutableBytesValue mutableBytes = bytes.mutableCopy();
+      mutableBytes.setBit(size, false);
+      finalBlank = mutableBytes.copy();
+    }
+
+    return new Bitlist(size, finalBlank, maxSize);
+  }
+
+  public static Bitlist of(int size, BytesValue bytes, long maxSize) {
+    return new Bitlist(size, bytes, maxSize);
+  }
+
+  public static Bitlist of(int maxSize) {
+    return new Bitlist(0, BytesValue.EMPTY, maxSize);
+  }
+
+  public static Bitlist of(int size, long bytes, long maxSize) {
+    UInt64 blank = UInt64.valueOf(bytes);
+    if (blank.getUsedBitCount() > size) {
+      throw new IndexOutOfBoundsException(
+          String.format("Input data %s exceeds Bitlist size of %s", bytes, size));
+    }
+
+    int neededBytes = (size + 7) / 8;
+    return new Bitlist(size, blank.toBytesValue().slice(0, neededBytes), maxSize);
+  }
+
+  private static BytesValue checkSize(BytesValue input, int size) {
+    int neededBytes = (size + 7) / 8;
+    if (neededBytes != input.size()) {
+      throw new IllegalArgumentException(
+          String.format(
+              "An attempt to initialize Bitlist with size %s using value %s with another size",
+              size, input));
+    }
+
+    return input;
+  }
+
+  public Bitlist setBit(int bitIndex, boolean bit) {
+    if (bitIndex > maxSize) {
+      throw new IndexOutOfBoundsException(
+          String.format("An attempt to set bit #%s for Bitlist with maxSize %s", bitIndex, size));
+    }
+    MutableBytesValue mutableCopy = mutableCopy();
+    mutableCopy.setBit(bitIndex, bit);
+    return new Bitlist(size, mutableCopy, maxSize);
+  }
+
+  public List<Integer> getBits() {
+    List<Integer> ret = new ArrayList<>();
+    for (int i = 0; i < size(); i++) {
+      if (getBit(i)) {
+        ret.add(i);
+      }
+    }
+    return ret;
+  }
+
+  public Bitlist and(Bitlist other) {
+    return Bitlist.of(Math.max(size, other.size), BytesValues.and(this.wrapped, other.wrapped), maxSize);
+  }
+
+  public Bitlist or(Bitlist other) {
+    return Bitlist.of(size, BytesValues.or(this.wrapped, other.wrapped), maxSize);
+  }
+
+  @Override
+  public String toString() {
+    StringBuilder ret = new StringBuilder("0b");
+    for (int i = 0; i < byteSize(); i++) {
+      ret.append(toBinaryString(get(i)));
+    }
+    return ret.toString();
+  }
+
+  @Override
+  public boolean equals(Object o) {
```

Comment:
`equals` and `hashCode` seems to be redundant here as they are already defined in `DelegatingBytesValue`.

Reply:
`DelegatingBytesValue` doesn't contain actual size and maxSize

Reply:
Got it!